### ISSUE TITLE
MNIST training hangs on small machine

### ISSUE THREAD

**Issue by jysohn23** [pytorch/xla#821]
MNIST training example hangs on smaller Client VM machines (ex. `n1-standard-1`). EAP customers have also reported this issue.

Repro instructions:
1. Create n1-standard-1 machine with PyTorch XLA image (or others if using docker).
2. Export XRT_TPU_CONFIG to point to TPU worker.
3. Run `python test/test_train_mnist.py` either with VM image installation or docker container and this will hang.

@zcain117 @taylanbil

**Comment by dlibenzi** [comment]
Thanks @jysohn23 
Any idea where it is hanging?
Dequeuing a new sample?
Somewhere in the C++ code?

**Comment by dlibenzi** [comment]
Setting sync-wait (wait=True) solves the issue:

https://github.com/pytorch/xla/blob/93b973dba3375705856510d10200fa4f1c41c571/torch_xla_py/xla_model.py#L326

But we can't do that.
I feel that's just a side effect though, the issue is somewhere else.

**Comment by dlibenzi** [comment]
To GDB on the client:
```
$ gdb --args python test/test_train_mnist.py --fake_data --log_steps=10000 --batch_size=125 
```

I use an high log step to avoid the loss.item() calls, and the batch size 125 to fit perfectly the MNIST sample size for 8 cores. It is not strictly needed.
When it gets stuck, CTRL^C and:
```
(gdb) set logging file /tmp/woo
(gdb) set logging on
Copying output to /tmp/woo.
(gdb) set pagination off
(gdb) thread apply all bt
```

Here is a log file:

https://gist.github.com/dlibenzi/81b2a778b09de2a7baabc08ad6631b79

The Python frames are Python threads.
But there is one C++ thread to us, on `xla::XrtComputationClient::TransferToServer()` and another one on `xla::XrtComputationClient::ExecuteComputation()`.
Pretty consistent.
No idea yet if on same device.

**Comment by dlibenzi** [comment]
It can be two things.
 1) The TPU computation hangs on the TPU VM side
 2) The GRPC response comes back but somehow the GRPC layer fails to deliver

I would not exclude `#2` as this is happening only with user VMs with low core/mem sizes.

**Comment by jysohn23** [comment]
I'm not how `xla::XrtComputationClient::TransferToServer()` was stuck on your run, but on my repro here is the client side logs: 

https://gist.github.com/jysohn23/2c249e3956653cc7d0b73a5301704d8c

On it we see that it's hanging on `xla::XrtComputationClient::ExecuteComputation()` but don't see `xla::XrtComputationClient::TransferToServer()` which makes a little more sense to me.

**Comment by jysohn23** [comment]
Also, did appear to have larger machines (`n1-standard-16`) hang on MNIST though much less frequently (1/10 runs) so this appears to be more broadly affecting.

**Comment by umgupta** [comment]
I tried mnist example on `n1-standard-4`, it hangs after 40 steps.

**Comment by jysohn23** [comment]
Can you try `n1-standard-16`? We haven't tested `n1-standard-4` but we are working on resolving the hanging issue. Thanks.

**Comment by umgupta** [comment]
@jysohn23 I can't try that. It is beyond what I can provision under free-tier. But I can confirm that things are fine with `n1-standard-8`. 

PS: Is there any good documentation on how to adapt cuda/cpu pytorch code to run on TPU? I am really having a hard time figuring it out. The examples as it is work but I am not sure what minimal changes I need to make my repository work.

**Comment by jysohn23** [comment]
> PS: Is there any good documentation on how to adapt cuda/cpu pytorch code to run on TPU? I am really having a hard time figuring it out. The examples as it is work but I am not sure what minimal changes I need to make my repository work.

We have this [API_GUIDE](https://github.com/pytorch/xla/blob/master/API_GUIDE.md) that should give you some guidance on how to migrate your APIs to make TPU compatible. Is this what you're looking for?

**Comment by umgupta** [comment]
Thanks! This is better. I was looking for a more elaborate tutorial(ish) readme but I will try to work with this. At least I now understand that `train_loop_fn` is not some magic. Since the examples use `run_tests` that confused me. I am not sure if it is a good idea. but if it is needed, I am willing to help with documentation.

**Comment by umgupta** [comment]
I also find that if you don't set `XRT_TPU_CONFIG` variable, the code hangs while configuring TPU worker at `grpc://localhost:51000`. I am not sure if that is the intended behavior so just wanted to point out.

**Comment by abhishekkrthakur** [comment]
The issue still seems to exist even on latest nightly. Is there a way to avoid getting stuck?

```
  File "train.py", line 358, in <module>
    model_parallel(train_loop_fn, train_data_loader)
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch_xla/distributed/data_parallel.py", line 159, in __call__
    thread.join()8s/it]
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/threading.py", line 1056, in join
    self._wait_for_tstate_lock()
  File "/anaconda3/envs/torch-xla-nightly/lib/python3.6/threading.py", line 1072, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
```

This happens _after_ all TPU cores have finished all batches of data for one epoch.

**Comment by dlibenzi** [comment]
Which `train.py` is that?

Threading based parallelism is also deprecated, and it is best to use multi-processing approach. Example:

https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py

**Comment by abhishekkrthakur** [comment]
@dlibenzi its my custom code based on https://github.com/pytorch/xla/blob/master/test/test_train_mnist.py

seems like https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py uses only one TPU core?

**Comment by dlibenzi** [comment]
No. The user deals only with one core, but the same code can scale up to a POD.

**Comment by abhishekkrthakur** [comment]
It seems like multiprocessing has other problems like pickling custom data loaders. I will create a new issue with code to reproduce soon.
